src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CoseKeyHelpers.cs

```diff
@@ -0,0 +1,261 @@
+
+            // begin writing CBOR object
+            writer.WriteStartMap();
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.Kty);
+            writer.WriteInt32((int)kty);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.Alg);
+            writer.WriteInt32((int)alg);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcCrv);
+            writer.WriteInt32((int)crv);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcX);
+            writer.WriteByteString(ecParams.Q.X);
+
+            WriteCoseKeyLabel(writer, CoseKeyLabel.EcY);
+            writer.WriteByteString(ecParams.Q.Y);
+
+            writer.WriteEndMap();
+
+            static (CoseKeyType kty, CoseCrvId crv) MapECCurveToCoseKtyAndCrv(ECCurve curve)
+            {
+                if (!curve.IsNamed)
+                {
+                    throw new ArgumentException("Only named curves supported in EC COSE keys.", nameof(curve));
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP256))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P256);
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP384))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P384);
+                }
+
+                if (MatchesOid(ECCurve.NamedCurves.nistP521))
+                {
+                    return (CoseKeyType.EC2, CoseCrvId.P521);
+                }
+
+                throw new ArgumentException("Unrecognized named curve", curve.Oid.Value);
+
+                bool MatchesOid(ECCurve namedCurve) => curve.Oid.Value == namedCurve.Oid.Value;
+            }
+
+            static CoseKeyAlgorithm MapHashAlgorithmNameToCoseKeyAlg(HashAlgorithmName name)
+            {
+                if (MatchesName(HashAlgorithmName.SHA256))
+                {
+                    return CoseKeyAlgorithm.ES256;
+                }
+
+                if (MatchesName(HashAlgorithmName.SHA384))
+                {
+                    return CoseKeyAlgorithm.ES384;
+                }
+
+                if (MatchesName(HashAlgorithmName.SHA512))
+                {
+                    return CoseKeyAlgorithm.ES512;
+                }
+
+                throw new ArgumentException("Unrecognized hash algorithm name.", nameof(HashAlgorithmName));
+
+                bool MatchesName(HashAlgorithmName candidate) => name.Name == candidate.Name;
+            }
+        }
+
+        public static (ECParameters, HashAlgorithmName) ReadECParametersAsCosePublicKey(CborReader reader)
+        {
+            Debug.Assert(reader.ConformanceLevel == CborConformanceLevel.Ctap2Canonical);
+
+            var ecParams = new ECParameters();
+
+            int? length = reader.ReadStartMap();
+            if (length != 5)
+            {
+                throw new FormatException("Unexpected number of elements in the CBOR cose key.");
+            }
+
+            // CTAP2 guarantees order of fields
```

Comment:
> I was curious to see if you were going to take advantage of this in the reader.

If you mean the field ordering guarantee, I think it's mostly effective in scenaria where you don't support optional fields. In the general case something [like this](https://gist.github.com/eiriktsarpalis/a6f7ab4b0c207c410b98dc9a1a2c12a2#file-cborcosekeyhelpers-cs-L40-L58) might be preferable.

Reply:
Oh, to be clear, I approve of taking the dependency; it's part of the niceness of the conformance mode :smile:.

Reply:
Sure, I suppose my point is that even assuming CTAP2 conformance mode, you might still need to opt for something like the original approach if you needed to tolerate arbitrary fields (e.g. because of forward compatibility concerns).